utilities/pulumi_state_splitter/pulumi_state_splitter/split.py

```diff
@@ -0,0 +1,177 @@
+
+    def to_state_file(self) -> pulumi_state_splitter.state_file.StateFile:
+        """Converts a split state to a Pulumi stack state file."""
+        return pulumi_state_splitter.state_file.StateFile(**self.model_dump())
+
+    def remove(self):
+        self._state_path.unlink()
+        if self.state.checkpoint.latest:
+            for resource in self.state.checkpoint.latest.resources:
+                path = self.path / self.resource_subpath(resource)
+                if resource.type == "pulumi:pulumi:Stack":
+                    (self.path / "outputs.yaml").unlink()
+                path.unlink()
+                pulumi_state_splitter.fs.rmdir_if_empty(path.parent)
+        for d in (self.path, self.path.parent):
+            pulumi_state_splitter.fs.rmdir_if_empty(d)
+
+    @classmethod
+    def find(
+        cls, backend_dir: pathlib.Path
+    ) -> Iterable[pulumi_state_splitter.stored_state.StackName]:
+        """Finds all directory states in the Pulumi backend directory."""
+        glob_states = cls._glob_all()._state_path
+        for state_path in backend_dir.glob(str(glob_states)):
+            stack_dir = state_path.parent
+            yield pulumi_state_splitter.stored_state.StackName(
+                project=stack_dir.parent.name,
+                stack=stack_dir.name,
+            )
+
+    def load(self):
+        """Loads the contents of the state directory."""
+        with self._state_path.open() as f:
+            data = yaml.load(f, yaml.Loader)
+        self.state = pulumi_state_splitter.model.State.model_validate(data)
+        if not self.state.checkpoint.latest:
+            return
+        resources = []
+        for dirpath, _, filenames in os.walk(self.path):
+            dirpath = pathlib.Path(dirpath)
+            if dirpath == self.path:
+                continue
+            for filename in filenames:
+                with (dirpath / filename).open() as f:
+                    data = yaml.load(f, yaml.Loader)
+                resource = pulumi_state_splitter.model.Resource.model_validate(data)
+                if resource.type == "pulumi:pulumi:Stack":
+                    with (self.path / "outputs.yaml").open() as f:
+                        resource.outputs = yaml.load(f, yaml.Loader)
+                resources.append(resource)
+        self.state.checkpoint.latest.resources = (
+            pulumi_state_splitter.model.Resource.find_parents(resources)
+        )
+
+    @classmethod
+    def resource_subpath(
+        cls, resource: pulumi_state_splitter.model.Resource
+    ) -> pathlib.Path:
+        """Determines where should a resource state be written to."""
+        if (
+            resource.parent_resource
+            and resource.parent_resource.type != "pulumi:pulumi:Stack"
+        ):
+            directory = cls.resource_subpath(resource.parent_resource).with_suffix("")
+        else:
+            directory = pathlib.Path()
+        return directory / resource.type / f"{resource.name}.yaml"
```

Comment:
Notice that `resource.type` will include `:` that will make this file/directory name illegal on MS Windows.

Reply:
And updated the unit test in 6c567c26976c945d7f4dffb698bab91e3c020eb4 .